### PR TITLE
Fix user queue in scheduler that was not thread-safe

### PR DESCRIPTION
**Pull request by justinjung04**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes an issue where user queue in scheduler wasn't thread-safe

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/scheduler/queue/user_queues_test.go [hunk 1]
```diff
@@ -457,6 +458,36 @@ func TestGetOrAddQueueShouldUpdateProperties(t *testing.T) {
 	}
 }
 
+func TestGetOrAddQueueConcurrency(t *testing.T) {
```
Comment (harry671003) [inline pkg/scheduler/queue/user_queues_test.go:461]: Is the test missing some assertions? Or was it just for testing for -race conditions?

Comment (justinjung04) [reply pkg/scheduler/queue/user_queues_test.go:461]: Yup, it's just to test with `-race` condition, which is part of the repo ci

**Review by harry671003 — APPROVED** [review]
LGTM. Thanks Justin.

File: pkg/scheduler/queue/user_queues.go [hunk 2]
```diff
@@ -103,6 +105,9 @@ func (q *queues) len() int {
 }
 
 func (q *queues) deleteQueue(userID string) {
+	q.userQueuesMx.Lock()
```
Comment (friedrichg) [inline pkg/scheduler/queue/user_queues.go:108]: this line is not covered by the unit test

File: pkg/scheduler/queue/user_queues.go [hunk 3]
```diff
@@ -237,6 +245,9 @@ func (q *queues) getNextQueueForQuerier(lastUserIndex int, querierID string) (us
 			continue
 		}
 
+		q.userQueuesMx.RLock()
```
Comment (friedrichg) [inline pkg/scheduler/queue/user_queues.go:248]: this line is not covered by the unit test

**Review by friedrichg — CHANGES_REQUESTED** [review]
Thank you, yes, there is a data race there. And I believe your proposed fix and test for the race are appropriate. I have only nits on covering all the changes with the test

File: pkg/scheduler/queue/user_queues.go (hunk 2 above)
Comment (justinjung04) [reply pkg/scheduler/queue/user_queues.go:108]: modified the test to cover this line

File: pkg/scheduler/queue/user_queues.go (hunk 3 above)
Comment (justinjung04) [reply pkg/scheduler/queue/user_queues.go:248]: modified the test to cover this line

**Review by danielblando — APPROVED** [review]

**Pull request merged**